homeassistant/components/launch_library/sensor.py

```diff
@@ -39,39 +43,75 @@ async def async_setup_platform(
     async_add_entities: AddEntitiesCallback,
     discovery_info: DiscoveryInfoType | None = None,
 ) -> None:
-    """Create the launch sensor."""
-    name = config[CONF_NAME]
-    session = async_get_clientsession(hass)
-    launches = PyLaunches(session)
+    """Import Launch Library configuration from yaml."""
+    _LOGGER.warning(
+        "Configuration of the launch_library platform in YAML is deprecated and will be "
+        "removed in Home Assistant 2022.4; Your existing configuration "
+        "has been imported into the UI automatically and can be safely removed "
+        "from your configuration.yaml file"
+    )
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": SOURCE_IMPORT},
+            data=config,
+        )
+    )
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the sensor platform."""
+
+    name = entry.data.get(CONF_NAME, DEFAULT_NAME)
+
+    coordinator = hass.data[DOMAIN]
 
-    async_add_entities([LaunchLibrarySensor(launches, name)], True)
+    async_add_entities(
+        [
+            NextLaunchSensor(coordinator, entry.entry_id, name),
+        ]
+    )
 
 
-class LaunchLibrarySensor(SensorEntity):
-    """Representation of a launch_library Sensor."""
+class NextLaunchSensor(CoordinatorEntity, SensorEntity):
+    """Representation of the next launch sensor."""
 
-    _attr_icon = "mdi:rocket"
+    _attr_attribution = ATTRIBUTION
+    _attr_icon = "mdi:rocket-launch"
 
-    def __init__(self, api: PyLaunches, name: str) -> None:
-        """Initialize the sensor."""
-        self.api = api
+    def __init__(
+        self, coordinator: DataUpdateCoordinator, entry_id: str, name: str
+    ) -> None:
+        """Initialize a Launch Library entity."""
+        super().__init__(coordinator)
         self._attr_name = name
+        self._attr_unique_id = f"{entry_id}_next_launch"
+
+    def get_next_launch(self) -> Launch | None:
+        """Return next launch."""
+        return next((launch for launch in self.coordinator.data), None)
+
+    @property
+    def native_value(self) -> str | None:
+        """Return the state of the sensor."""
+        if next_launch := self.get_next_launch():
+            self._attr_available = True
```

Comment:
This won't work, as the `native_value` property isn't called then the sensor is unavailable.

As this is the coordinator, you can better implement a method that updates the state each time the update coordinator signals.

e.g.:

```py
class NextLaunchSensor(CoordinatorEntity, SensorEntity):
    """Representation of the next launch sensor."""

    _attr_attribution = ATTRIBUTION
    _attr_icon = "mdi:rocket-launch"
    _next_launch: Launch | None = None

    def __init__(
        self, coordinator: DataUpdateCoordinator, entry_id: str, name: str
    ) -> None:
        """Initialize a Launch Library entity."""
        super().__init__(coordinator)
        self._attr_name = name
        self._attr_unique_id = f"{entry_id}_next_launch"

    @property
    def native_value(self) -> str | None:
        """Return the state of the sensor."""
        if self._next_launch is None:
            return None
        return self._next_launch.name

    @property
    def available(self) -> bool:
        """Return if the sensor is available."""
        return super().available and self._next_launch is not None

    @property
    def extra_state_attributes(self) -> dict[str, Any] | None:
        """Return the attributes of the sensor."""
        if self._next_launch is None:
            return None
        return {
            ATTR_LAUNCH_TIME: self._next_launch.net,
            ATTR_AGENCY: self._next_launch.launch_service_provider.name,
            ATTR_AGENCY_COUNTRY_CODE: self._next_launch.pad.location.country_code,
            ATTR_STREAM: self._next_launch.webcast_live,
        }

    @callback
    def _handle_coordinator_update(self) -> None:
        """Handle updated data from the coordinator."""
        self._next_launch = next((launch for launch in self.coordinator.data), None)
        super()._handle_coordinator_update()

    async def async_added_to_hass(self) -> None:
        """When entity is added to hass."""
        await super().async_added_to_hass()
        self._handle_coordinator_update()
```

Reply:
Right, just so I understand it. `_handle_coordinator_update` is called everytime by the coordinator when it does an update?